### DISCUSSION TITLE
useFieldArray resets/unregisters form field manually registered after deleting, moving ...

### DISCUSSION THREAD

**Discussion by jlame** [react-hook-form/react-hook-form#3218] [Q&A] [answered]
**Describe the bug**
I'm using some complex component in my form which i register manually and setValue onChange. When using delete, move ... from fieldArray, the value are removed from form and even worst seems like my component are not registered anymore.
In my example, the "complex component is replaced by the **Test** component.

Maybe i'm doing something wrong or i forgot to do something (like i need to register it each time ? but how do i do that without ending in infinite loop with my useEffect ?), i'd really appreciate help on that one. 
Thank you ! 😄 

**EDIT:** Managed to make something work in this one : https://codesandbox.io/s/react-hook-form-usefieldarray-forked-w5hri,
i had to : 

1. Add fields as dependency of my component useEffect in order to make it register again after a remove, move ...
2. Add a watcher to keep the current value so that the useEffect use it (as a default value) when registering back

**This seems pretty heavy to me, maybe there's another way to make it work, but didn't find it yet ... and i still have some issue when moving because of the watch where it keeps the previous value, so this is not viable too**

**To Reproduce**

1. Go to https://codesandbox.io/s/react-hook-form-usefieldarray-forked-mvl2v
2. Click on append
3. Submit and see that the "test" value is in the form
4. Delete one item
5. Submit and see that the "test" value has been removed from the form

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-mvl2v

**Expected behavior**
The value should still be there after removing, moving etc...

**Screenshots**
![image](https://user-images.githubusercontent.com/14272767/96581401-0d492500-12da-11eb-91ea-ee978ed6254a.png)

**Desktop (please complete the following information):**

- windows 10
- Browser chrome
- Version 86

**Additional context**

**Comment by DamodarSojka** [comment]
There is definitely something wrong with the useFieldArray even in a very basic case.
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-bemv3?file=/src/index.js
1. click append 
2. click prepend

GetValues returns an empty object as seen in the {JSON.stringify(getValues())}.

**Comment by bluebill1049** [comment]
It's always better to give `inputs` to hook form at useFieldArray.

- I would turn those custom register into hidden inputs and give a try. 
- set `shouldUnregister: false`

**Comment by DamodarSojka** [comment]
Did you have a look at my comment? It's not really the same issue but I though it's related and therefore didn't create another issue for it.

**Comment by bluebill1049** [comment]
> Did you have a look at my comment? It's not really the same issue but I though it's related and therefore didn't create another issue for it.

Question: why use `getValues()` in the render? it's always going to be one render behind the real value of the form.

<img width="579" alt="Screen Shot 2020-10-21 at 8 43 10 am" src="https://user-images.githubusercontent.com/10513364/96647486-7a5fc780-1379-11eb-8272-1e0dc3ff4fca.png">

**Comment by bluebill1049** [comment]
@DamodarSojka `getValues()` is made so you can read what's in the form, call it during the render will cause the issue. eg:

1. getValue()
2. mount: <input />
3. mount: <input />

it will potentially return empty form because 2 and 3 are mounted after.

**Comment by bluebill1049** [comment]
> Did you have a look at my comment? It's not really the same issue but I though it's related and therefore didn't create another issue for it.

Also, make sure you run the latest in the CSB :)

**Comment by DamodarSojka** [comment]
> Question: why use `getValues()` in the render? it's always going to be one render behind the real value of the form.

I am passing the values to another component.
I think it also caused issues in a onBlur callback but I would need to double check that.
What would you recommend instead?

> Also, make sure you run the latest in the CSB :)

Somehow I assumed the link I forked was already latest, I also did notice issues about getValues in my project anyway (but it seems the issues are justified).

**Comment by bluebill1049** [comment]
> I am passing the values to another component.
> I think it also caused issues in a onBlur callback but I would need to double check that.
> What would you recommend instead?

I need more context.

**Comment by jlame** [comment]
> It's always better to give `inputs` to hook form at useFieldArray.
> 
> * I would turn those custom register into hidden inputs and give a try.
> * set `shouldUnregister: false`

Thank you for your reply

I just tried that using the hidden input method and the issue i bump into is that when from my complex component i try to use setValue with an object, this object is cast into a string (i guess because the register element is an input) and so i end up with "[object Object]" as a value. Is there a way to keep the object ? It's also an issue for the defaultValue where i'm trying to put an object to the input default value ...

Because one of my complex component will output a complex object structure which i need to keep.

To give you more context i have something like that : `
       <input
            type="hidden"
            name={name}
            defaultValue={defaultValue}
            required
            ref={register(rules)}
        />
        <DateRangePicker
            onChange={(event, value) => {
                setValue(name, {
                    startDate: value.startDate,
                    endDate: value.endDate,
                })
            }}
            endDate={defaultValue?.endDate}
            startDate={defaultValue?.startDate}
        />
`

**Comment by bluebill1049** [comment]
> > It's always better to give `inputs` to hook form at useFieldArray.
> > 
> > * I would turn those custom register into hidden inputs and give a try.
> > * set `shouldUnregister: false`
> 
> Thank you for your reply
> 
> I just tried that using the hidden input method and the issue i bump into is that when from my complex component i try to use setValue with an object, this object is cast into a string (i guess because the register element is an input) and so i end up with "[object Object]" as a value. Is there a way to keep the object ? Because one of my complex component will output a complex object structure which i need to keep.

if you set `shouldUnregister: false` you should be able to setValue with `object`, provide a codesandbox if not working out.

**Comment by DamodarSojka** [comment]
> I need more context.

Passing value to other component - I have an array of dropdowns, user can control the number of said dropdowns, picked values can't be duplicated so each dropdown has to know what other values were already picked.

onBlur callback - onBlur event of said dropdowns is used to add empty dropdown if the blured dropdown was last and just got a value, or remove a dropdown if it blured, isn't last in the array, there are at least two other dropdowns and it's empty.

I hope this time it's not too much context!
Let me know if you need any more information.

**Comment by jlame** [comment]
> > > It's always better to give `inputs` to hook form at useFieldArray.
> > > 
> > > * I would turn those custom register into hidden inputs and give a try.
> > > * set `shouldUnregister: false`
> > 
> > 
> > Thank you for your reply
> > I just tried that using the hidden input method and the issue i bump into is that when from my complex component i try to use setValue with an object, this object is cast into a string (i guess because the register element is an input) and so i end up with "[object Object]" as a value. Is there a way to keep the object ? Because one of my complex component will output a complex object structure which i need to keep.
> 
> if you set `shouldUnregister: false` you should be able to setValue with `object`, provide a codesandbox if not working out.

So i've managed to set an object using the hidden input method, but i still have an issue when removing, moving items : 
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-s2hvs

The key is still there in the data object, but the value is set to undefined when doing one of the above action.

**Comment by bluebill1049** [comment]
> > I need more context.
> 
> Passing value to other component - I have an array of dropdowns, user can control the number of said dropdowns, picked values can't be duplicated so each dropdown has to know what other values were already picked.
> 
> onBlur callback - onBlur event of said dropdowns is used to add empty dropdown if the blured dropdown was last and just got a value, or remove a dropdown if it blured, isn't last in the array, there are at least two other dropdowns and it's empty.
> 
> I hope this time it's not too much context!
> Let me know if you need any more information.

Can you produce a simple codesandbox for what you try to achieve?

**Reply by DamodarSojka** [reply]
Ok, I got time to sit to that again, so for the using getValues in the render phase:
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-wpsje?file=/src/index.js

Essentially there is a set of options the user can pick and three selects which user has to pick value for, already picked values are filtered out from the set of available options for each of the selects.

Let me know if it makes sense and if I can provide any more examples.

**Reply by bluebill1049** [reply]
I think `watch` is the better option here for what you try to achieve.

**Reply by DamodarSojka** [reply]
Could you elaborate on how you would approach using watch for this? 
Would you say it would be a general approach to have watch use say useState to save portions of the form data to be used within render?

**Reply by bluebill1049** [reply]
if you want to subscribe to form inputs, you can either use `watch` which render at the root level or use `useWatch` to watch any part of the FieldArray and trigger isolated re-render at your component level.

**Comment by bluebill1049** [comment]
> > > > It's always better to give `inputs` to hook form at useFieldArray.
> > > > 
> > > > * I would turn those custom register into hidden inputs and give a try.
> > > > * set `shouldUnregister: false`
> > > 
> > > 
> > > Thank you for your reply
> > > I just tried that using the hidden input method and the issue i bump into is that when from my complex component i try to use setValue with an object, this object is cast into a string (i guess because the register element is an input) and so i end up with "[object Object]" as a value. Is there a way to keep the object ? Because one of my complex component will output a complex object structure which i need to keep.
> > 
> > 
> > if you set `shouldUnregister: false` you should be able to setValue with `object`, provide a codesandbox if not working out.
> 
> So i've managed to set an object using the hidden input method, but i still have an issue when removing, moving items :
> https://codesandbox.io/s/react-hook-form-usefieldarray-forked-s2hvs
> 
> The key is still there in the data object, but the value is set to undefined when doing one of the above action.

I don't think it's going to work with the way you have provided above. here is what I would do:

I would create a separate useFieldArray in the `<Test />`. -> https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-x7btr as far as i know, you are trying to achieve nested array field.

or : https://codesandbox.io/s/react-hook-form-usefieldarray-forked-0r94h (I am not recommending this way tho)

**Comment by jlame** [comment]
Hmm, in fact i only have one array which is my items and these items come with simple field (input, select) but also more complex one. So the output object i try to have is something like : [
      { atext: "something", aNumber: 1, aDate: { startDate: new Date(), endDate: new Date() },  ...},
      { atext: "another thing", aNumber: 2, aDate: { startDate: new Date(), endDate: new Date() },  ...}
]

From my point of view, this is not nested array, am i wrong ?

**Comment by bluebill1049** [comment]
> Hmm, in fact i only have one array which is my items and these items come with simple field (input, select) but also more complex one. So the output object i try to have is something like : [
> { atext: "something", aNumber: 1, aDate: { startDate: new Date(), endDate: new Date() }, ...},
> { atext: "another thing", aNumber: 2, aDate: { startDate: new Date(), endDate: new Date() }, ...}
> ]
> 
> From my point of view, this is not nested array, am i wrong ?

That's not, then you should two inputs
```
<input name="test[0].aDate.startDate />
<input name="test[0].aDate. endDate />
```

**Reply by jlame** [reply]
Yes for the date range case i can do that, but for other components it will probably get painful to add an input for each field of an object
> or : https://codesandbox.io/s/react-hook-form-usefieldarray-forked-0r94h (I am not recommending this way tho)

Can you tell me why don't you recommend that one ? In fact that  is what i was doing at the very begining before trying the hidden input method. 
I guess that's because with shouldUnregister to false, i might not have a "clean" data object as conditionnal field would be there as long as they displayed once and that's also forces me to implement more complex validation function for conditionnal field ... Am i right ?

**Reply by bluebill1049** [reply]
> I guess that's because with shouldUnregister to false, i might not have a "clean" data object as conditionnal field would be there as long as they displayed once and that's also forces me to implement more complex validation function for conditionnal field ... Am i right ?

yes, it's always recommended to have inputs after all hook form is working better with inputs.

**Comment by jlame** [comment, marked as answer]
So right now using the not recommanded option (how it works suits better how i'm going to manage complex component), i managed to make almost everything work ... The only thing left is that after removing, moving etc ..., the validation function i pass doesnt work anymore (it does not get executed anymore after one of the operations).
here's a sandbox : 
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-s2hvs?file=/src/index.js

1. click on validate : a required error is present and the validate function was executed
2. append one and click on validate : still fine
3. remove the appended item
4. validate : the form end up in error with an empty object and the validate function was not executed
5. the form is stuck on that empty error, and i need to use clearErrors() to unblock it BUT validation is not done anymore as if it is not registered anymore

do you think i can work around this ?

Thanks for your time, really sorry to bother you but the usecase i'm facing makes it almost impossible to do it in another manner i guess ...

**Reply by bluebill1049** [reply]
Thanks for the detailed codesandbox and description. I will take a look at it today.

**Reply by bluebill1049** [reply]
> 1. click on validate : a required error is present and the validate function was executed

<img width="453" alt="Screen Shot 2020-10-22 at 1 13 03 pm" src="https://user-images.githubusercontent.com/10513364/96813835-5333f380-1468-11eb-8af3-dc54f810f9ff.png">

did you forget to save? there is no validate to click.

**Reply by bluebill1049** [reply]
`rhf.setValue(`test[${index}].test`, { a: "aaaaa", b: "bbbbbb" });` is it because you are missing `{ shouldValidate: true }` the third argument.

**Reply by bluebill1049** [reply]
also, there is another option (if this method is not working out for you), build a Controlled component that host all your `data`, and wrapped with `Controller`. (but let's focus on what you got right now)

**Reply by jlame** [reply]
sorry what i meant by "validate" was the submit button haha ... so basically at the begining my sandbox is built so that you have error on submit, by clicking on the setValue button you correct the error.
The issue is that if you don't correct the error and use remove, move etc... It just wreck the whole form => The form is always in error (with an empty object as errors) and even clicking on the button to set the value doesn't make it good again.
The only solution is to use clearErrors, but i've noticed that the rule i've registered doesn't work anymore (not executed anymore).
I've tried to add the shouldValidate: true, but that's doesn't change anything, i guess the form at this point is in a strange state and there might be a little bug somewhere

**Reply by jlame** [reply]
![empty-error](https://user-images.githubusercontent.com/14272767/96839062-f248f300-1448-11eb-865c-b5d431352d1d.gif)

**Reply by jlame** [reply]
the item i'm removing above is the one without error (after i use the setValue on it), so i should still have the first item error, but end up with an empty object and the validate function is not executed anymore.

**Edit :** Meanwhile i'm tryng the controller because my date range picker is a controlled component but i have other which are uncontrolled ... and that would need to make a controlled wrapper then use controller, i would have prefer for it to work directly with rhf though :)

**Reply by bluebill1049** [reply]
looks like you didn't give `button` type to your button. (not related to the problem)

**Reply by bluebill1049** [reply]
Can you test again with this CSB? I think i have found an issue: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-156ms?file=/src/index.js

**Reply by jlame** [reply]
ok so now, after deleting and submiting again, i still have the error and not an empty object anymore. **But** the validate function is not executed when submitting which i guess is not normal cause even if i use the setValue it won't revalidate and so stay in error forever.
(BTW what do you mean by CSB ?) 

Thanks again for your time, i appreciate :)

**Reply by bluebill1049** [reply]
no problem mate, because setValue need `shouldValidate`

**Reply by bluebill1049** [reply]
try again: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-156ms?file=/src/index.js:694-708

**Reply by jlame** [reply]
already tried it, still not okay, anyway the submit launches validation so that shouldn't be needed i guess. But even with it, what i see is that the validate function is not executed. It seems like the rule was unregistered or something like that, like it doesn't exist anymore (i've left a console log in the function and it's not displayed).

**Reply by bluebill1049** [reply]
aha, that's why I wasn't recommended doing the custom register at useEffect with useFieldArray, hook form useFieldArray doing the shuffle and input re-registration internally. try this hacky approach: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-156ms?file=/src/index.js

**Reply by bluebill1049** [reply]
nvm doesn't work correctly...

**Reply by bluebill1049** [reply]
I will try to convert into `Controller` for you.

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-156ms?file=/src/index.js

**Reply by jlame** [reply]
yeah i also made my date range work with controller, but i still need the shouldUnregister: false flag or else i lose my values on move,remove etc... 
I guess i don't have another choice right now, only way to make it work is to use controller with shouldUnregister and controlled component right ?

**Reply by bluebill1049** [reply]
you don't need `shouldUnregister false`, please test my CSB above.

**Reply by jlame** [reply]
indeed, i thought i did the test and saw that it was needed but no its not :) 
ok then i'll have to go with full controller for all my component which are not simple input.

Thanks a lot for your time Bill !

**Reply by bluebill1049** [reply]
No worries, buddy. Let me know if you still encounter problems, happy to help everyone.

**Reply by jlame** [reply]
Hi @bluebill1049 , currently encountering the empty object errors again,  but i can't reproduce it in a sandbox :( ...
In fact, i just have a conditional input like { isTrue && (<mycomponent>)} in that component (form is using shouldunregister: false)  :  

I registered a required rule on a simple input, when submitting i have the error. Then when changing isTrue to make my component go away, if i try to submitting again i get that empty error object {} and that just blocks the whole form.

I don't understand in which case and why i get this empty object as an error, can you explain me in which case that happens ? Knowing that, if i change my isTrue value back to display the component, the error reappears (so it gets back to normal).

**Reply by bluebill1049** [reply]
sounds really strange. it could be a bug, i am going to release a minor version today, not sure if this will resolve your problem, worth to give a shot otherwise you will need to help me to reproduce the problem in CSB.

**Reply by jlame** [reply]
> sounds really strange. it could be a bug, i am going to release a minor version today, not sure if this will resolve your problem, worth to give a shot otherwise you will need to help me to reproduce the problem in CSB.

thanks i will try to make a csb today

**Comment by alextaymx** [comment]
i want to achieve array of field array

izit possible to set all their values from outside? 

my idea is: 
setValue('test', [  {key:"a", value:  [  {key:"b"}  ]  }  ]   )
somehow, the setValue only works for the first layer of field array, nested field array "value" in this case is not set

am i doing it wrong?

**Reply by bluebill1049** [reply]
need more detail, which version and supply a codesandbox?

**Reply by NairiAreg** [reply]
I have the same error
